---
title: "Analysis: Average gene expression of LR pairs"
author: "Joan Kant"
params:
    input_file: "/Users/joankant/Desktop/gaitigroup/Users/Joan/scrnaseq-cellcomm/output/CCI_CellClass_L2_2_reassigned_samples/402_aggregation/402c_aggregation_integration.rds"
    output_dir: "/Users/joankant/Desktop/gaitigroup/Users/Joan/scrnaseq-cellcomm/002_WIP_FIGURES_local/scRNAseq/CCI_CellClass_L2_2_reassigned_samples"
    unique_interactions: "/Users/joankant/Library/CloudStorage/OneDrive-UHN/Spatial_GBM/Analysis/WIP/scRNAseq/CCI/CCI_CellClass_L2_2_reassigned_samples_unique_interactions_neuron_invasive_high.xlsx"
    condition_varname: "Region"
    meta: "/Users/joankant/Desktop/gaitigroup/Users/Joan/scrnaseq-cellcomm/output/CCI_CellClass_L2_2_reassigned_samples/000_data/gbm_regional_study__metadata.rds"
    condition_oi: "PT"
    source: "Neuron"
    target: "Invasive-high OPC/NPC1"
    interactions_db: "/Users/joankant/Desktop/gaitigroup/Users/Joan/scrnaseq-cellcomm/data/interactions_db/ref_db.rds"
    avg_expr: "/Users/joankant/Desktop/gaitigroup/Users/Joan/scrnaseq-cellcomm/output/average_expression_and_pseudobulk/average_expression_by_Sample_CCI_CellClass_L2_2.rds"
    cutoff: 3
    top_n: 10
output: html_document
---

```{r ,echo = FALSE, include = FALSE}
knitr::opts_chunk$set(echo = FALSE, out.height = "\\textheight", out.width = "\\textwidth", warning = FALSE, message = FALSE)

# # TODO remove later, just for testing temporary
library(GaitiLabUtils)
library(GBMutils)
set_wd()

# Set up logging
logr <- init_logging(log_level = 5)

# Load **libraries**
pacman::p_load(glue, data.table, tidyverse, stringr)
devtools::load_all("./", export_all = FALSE)

# Load additional libraries
pacman::p_load(ggplot2, ggpubr, scales, ggtext, ComplexHeatmap, ggrepel, readxl, ggrastr)

# TMP: only temporary for testing (interactively)
# params <- list()
# params$input_file <- "/Users/joankant/Desktop/gaitigroup/Users/Joan/scrnaseq-cellcomm/output/CCI_CellClass_L2_2_reassigned_samples/402_aggregation/402c_aggregation_integration.rds"
# params$output_dir <- "/Users/joankant/Desktop/gaitigroup/Users/Joan/scrnaseq-cellcomm/002_WIP_FIGURES_local/scRNAseq/CCI_CellClass_L2_2_reassigned_samples"
# params$unique_interactions <- "/Users/joankant/Library/CloudStorage/OneDrive-UHN/Spatial_GBM/Analysis/WIP/scRNAseq/CCI/CCI_CellClass_L2_2_reassigned_samples_unique_interactions_neuron_invasive_high.xlsx"
# params$condition_varname <- "Region"
# params$source <- "Neuron"
# params$target <- "Invasive-high OPC/NPC1"
# params$condition_oi <- "PT"
# params$meta <- "/Users/joankant/Desktop/gaitigroup/Users/Joan/scrnaseq-cellcomm/output/CCI_CellClass_L2_2_reassigned_samples/000_data/gbm_regional_study__metadata.rds"
# params$interactions_db <- "/Users/joankant/Desktop/gaitigroup/Users/Joan/scrnaseq-cellcomm/data/interactions_db/ref_db.rds"

# params$avg_expr <- "/Users/joankant/Desktop/gaitigroup/Users/Joan/scrnaseq-cellcomm/output/average_expression_and_pseudobulk/average_expression_by_Sample_CCI_CellClass_L2_2.rds"
# params$cutoff <- 3

# #log_info("Create output directory...")
create_dir(params$output_dir)


# Setup for visualization
options(ggrepel.max.overlaps = Inf)
gene_labels <- c("background", "ligand", "ligand/receptor", "receptor")
# For filtering later on
gene_labels_orderby <- c("ligand/receptor", "ligand", "receptor", "background")
label_color <- GBMutils::load_color_palette("General_groups")[1:4]
names(label_color) <- gene_labels

# Remove '/' for saving figures
source_target_oi <- paste0(params$source, "__", params$target)
source_target_oi_formatted <- str_replace_all(source_target_oi, c("/" = "_", " " = "_"))

target_mean <- paste0("mean_", params$target)
target_sd <- paste0("sd_", params$target)
source_mean <- paste0("mean_", params$source)
source_sd <- paste0("sd_", params$source)

get_genes_from_unique_interactions <- function(unique_interactions_path, source_oi, target_oi, condition_varname, condition_oi, is_directed = TRUE, top_n = NULL) {
    st <- paste0(source_oi, "__", target_oi)
    st_rev <- paste0(target_oi, "__", source_oi)

    if (is_directed) {
        st_vec <- c(st)
    } else {
        st_vec <- c(st, st_rev)
    }

    unique_interactions <- read_excel(unique_interactions_path) %>%
        filter(source_target %in% st_vec, !!sym(condition_varname) == condition_oi) %>%
        mutate(log10p = -log10(pval)) %>%
        arrange(desc(log10p), desc(CellChat_score), .by_group = TRUE) %>%
        distinct(complex_interaction, .keep_all = TRUE) %>%
        # Ranking
        mutate(unique_order_id = min_rank(desc(log10p))) %>%
        select(unique_order_id, complex_interaction, unique_order_id) %>%
        ungroup()
    # Add dummy, used for merging later, and format interaction string for visualization
    top_n <- ifelse(is.null(top_n), nrow(unique_interactions), top_n)
    unique_interactions <- unique_interactions %>%
        filter(unique_order_id <= top_n) %>%
        pull(complex_interaction)
    return(unique_interactions)
}

generate_scatter_plots <- function(
    expr_df,
    detected_interactions_path,
    condition_oi,
    condition_varname,
    user_input_list,
    source_oi,
    target_oi,
    cutoff = 3,
    is_directed = FALSE) {
    source_target_oi <- paste0(source_oi, "__", target_oi)

    st <- paste0(source_oi, "__", target_oi)
    st_rev <- paste0(target_oi, "__", source_oi)
    st_formatted <- str_replace_all(st, c("/" = "_", " " = "_"))


    if (is_directed) {
        st_vec <- c(st)
    } else {
        st_vec <- c(st, st_rev)
    }


    # Load interactions + basic filtering: source-target of interest + condition of interest
    detected_interactions <- readRDS(detected_interactions_path) %>%
        filter(
            pval < 0.05, source_target %in% st_vec,
            !!sym(condition_varname) == condition_oi,
            # Either
            # (1) keep interactions with genes of interest OR
            # (2) keep interactions from list of interactions of interest (e.g. unique interactions)
            (str_detect(complex_interaction, paste0(user_input_list, collapse = "|"))) |
                (complex_interaction %in% user_input_list)
        ) %>%
        separate(source_target, into = c("source", "target"), sep = "__") %>%
        left_join(interactions_db %>% select(complex_interaction, ligand_complex, receptor_complex)) %>%
        ungroup() %>%
        mutate(interaction_oi = TRUE)

    # Obtain genes expressed by source (incl. subunits)
    source_ligand <- detected_interactions %>%
        select(-target, -receptor_complex) %>%
        separate_rows(ligand_complex, sep = "\\:") %>%
        rename(gene = ligand_complex, label = source) %>%
        mutate(role_in_interaction = "ligand")

    # Obtain genes express by target (incl. subunits)
    target_receptor <- detected_interactions %>%
        select(-source, -ligand_complex) %>%
        separate_rows(receptor_complex, sep = "\\:") %>%
        rename(gene = receptor_complex, label = target) %>%
        mutate(role_in_interaction = "receptor")

    # Combine in one dataframe w/ all genes involved in interactions of interest and the cell type that expresses it
    label_gene_combi <- rbind(source_ligand, target_receptor) %>%
        # select(-complex_interaction) %>%
        select(!!sym(condition_varname), gene, label, role_in_interaction, interaction_oi) %>%
        distinct()

    # Expression for condition of interest
    df_region <- expr_df %>%
        # Select condition of interest
        filter(!!sym(condition_varname) == condition_oi) %>%
        # Add:
        # - 'label' - either 'source' (source_oi),'target' (target_oi) or 'NA'
        # - 'interaction_oi' - marks whether the gene is part of an interaction of interest
        left_join(label_gene_combi, by = c(condition_varname, "gene")) %>%
        # If gene is not part of interactions of interest, set to FALSE
        mutate(interaction_oi = ifelse(is.na(interaction_oi), FALSE, interaction_oi)) %>%
        # TODO Remove filter later, just for testing
        # filter(interaction_oi) %>%
        # Remove duplicates: genes that are expressed by both source_oi AND target_oi (e.g. both a 'ligand' and 'receptor')
        arrange(gene, category) %>%
        distinct(gene, .keep_all = TRUE) %>%
        select(interaction_oi, gene, category, !!sym(condition_varname), !!sym(target_mean), !!sym(source_mean)) %>%
        ungroup()

    # Modelling LOESS based on source/target expression
    loess_fit <- loess(df_region %>% pull(!!sym(source_mean)) ~ df_region %>% pull(!!sym(target_mean)))

    # # Take residuals
    resid <- scale(residuals(loess_fit), scale = TRUE, center = TRUE)

    df_region[!(is.na(df_region[source_mean]) | is.na(df_region[target_mean])), "resid"] <- resid
    df_region[!(is.na(df_region[source_mean]) | is.na(df_region[target_mean])), "fitted"] <- loess_fit$fitted


    df_region <- df_region %>%
        ungroup() %>%
        mutate(resid = ifelse(abs(resid) > cutoff, sign(resid) * cutoff, resid), )

    # Change factor levels order
    df_region <- df_region %>% mutate(category = factor(as.character(category), levels = gene_labels))

    # Plot with LOESS
    p_with_loess <- ggplot(data = df_region, aes(x = !!sym(source_mean), y = !!sym(target_mean), label = gene)) +
        geom_point_rast(size = 0.1, aes(color = resid)) +
        labs(
            x = paste0("log<sub>2</sub>(exp<sub>avg</sub> + 1) in ", source_oi),
            y = paste0("log<sub>2</sub>(exp<sub>avg</sub> + 1) in ", target_oi),
            title = condition_oi
        ) +
        scale_color_gradient2(
            midpoint = 0, limits = c(-cutoff, cutoff),
            low = label_color[4],
            high = label_color[2],
            guide = guide_colorbar(
                title = "Scaled residuals", barwidth = unit(10, "lines"), barheight = unit(0.7, "lines"),
            )
        ) +
        GBM_theme() +
        ggnewscale::new_scale("colour") +
        scale_color_manual(
            values = label_color,
            guide = guide_legend(
                override.aes = list(size = 7, alpha = 1), title = "Label"
            )
        ) +
        # Plot genes that are part of interactions of interest and label them (make them bigger in size to stand out)
        ggrepel::geom_label_repel(
            data = df_region %>% filter(interaction_oi), aes(label = ifelse(category != "background", gene, ""), color = category), show.legend = FALSE,
            min.segment.length = unit(0.1, "lines"),
        ) +
        geom_point(data = df_region %>% filter(interaction_oi), alpha = 1, aes(color = category)) +
        scale_y_continuous(breaks = scales::pretty_breaks()) +
        scale_x_continuous(breaks = scales::pretty_breaks())

    # Plot without LOESS
    p <- ggplot(data = df_region, levels = gene_labels, aes(x = !!sym(source_mean), y = !!sym(target_mean), label = gene)) +
        # Only add genes that are not of interest, set category to 'background' (1st layer)
        geom_point_rast(
            data = df_region %>% filter(!interaction_oi) %>% mutate(category = "background"),
            aes(color = category), size = 0.1
        ) +
        labs(
            x = paste0("log<sub>2</sub>(exp<sub>avg</sub> + 1) in ", source_oi),
            y = paste0("log<sub>2</sub>(exp<sub>avg</sub> + 1) in ", target_oi),
            title = condition_oi
        ) +
        scale_color_manual(
            values = label_color,
            guide = "none"
        ) +
        GBM_theme() +
        ggnewscale::new_scale("colour") +
        scale_color_manual(
            values = label_color,
            guide = guide_legend(
                override.aes = list(size = 7, alpha = 1), title = "Label"
            )
        ) +
        # Plot genes that are part of interactions of interest and label them (2nd layer)
        geom_point_rast(data = df_region %>% filter(interaction_oi), aes(color = category)) +
        ggrepel::geom_label_repel(
            data = df_region %>% filter(interaction_oi), aes(label = gene, color = category),
            show.legend = FALSE, min.segment.length = unit(0.1, "lines"),
        ) +
        scale_y_continuous(breaks = scales::pretty_breaks()) +
        scale_x_continuous(breaks = scales::pretty_breaks())

    return(list(interactions_oi = detected_interactions, p_loess = p_with_loess, p = p))
}
```

```{r, interactions_db_prep, echo = FALSE}
# Load interactions db
interactions_db <- readRDS(params$interactions_db)

# Extract ligands/receptors (splitting complexes)
all_ligands <- interactions_db$ligand_complex %>%
    str_split("\\:") %>%
    unlist() %>%
    unique()
all_receptors <- interactions_db$receptor_complex %>%
    str_split("\\:") %>%
    unlist() %>%
    unique()
genes_in_db <- c(all_ligands, all_receptors) %>% unique()
# Genes labeled as ligand AND receptor
all_dual <- intersect(all_ligands, all_receptors) %>% unique()
```

## Prepare gene expression 
Formatting gene expression and labeling genes based on interactions database. A gene can be labeld as: 'ligand', 'receptor', 'ligand/receptor' or 'background' 
```{r, gene_expr_prep, echo = TRUE }
# Get df of sample + region label
meta <- readRDS(params$meta) %>%
    select(Sample, !!sym(params$condition_varname)) %>%
    remove_rownames() %>%
    distinct()

# Load average expression
avg_exp <- readRDS(params$avg_expr)

colnames_split <- str_split(names(colnames(avg_exp)), ":|]", simplify = TRUE)[, c(2, 4)]
new_colnames <- apply(colnames_split, 1, paste0, collapse = ":")
colnames(avg_exp) <- new_colnames

# Average gene expression by condition, i.e. region
avg_exp_long <- data.frame(avg_exp, check.names = FALSE) %>%
    rownames_to_column("gene") %>%
    pivot_longer(names_sep = ":", names_to = c("label", "Sample"), cols = all_of(colnames(avg_exp)), values_to = "avg_expression") %>%
    left_join(meta) %>%
    group_by(!!sym(params$condition_varname), label, gene) %>%
    summarise(sd = sd(avg_expression), mean = mean(avg_expression)) %>%
    mutate(Region = factor(!!sym(params$condition_varname), levels = names(GBMutils::load_color_palette("Region")))) %>%
    ungroup()
head(avg_exp_long)

# Average expression (and SD) for source-target of interest (wide-format)
df <- avg_exp_long %>%
    # Log-transform
    mutate(sd = log2(sd + 1), mean = log2(mean + 1)) %>%
    filter(label %in% c(params$source, params$target)) %>%
    pivot_wider(names_from = label, values_from = c(mean, sd), values_fill = NA) %>%
    rowwise() %>%
    # Label the genes based on the interactions database
    mutate(category = factor(case_when(
        (gene %in% all_dual) ~ "ligand/receptor",
        gene %in% all_ligands ~ "ligand",
        gene %in% all_receptors ~ "receptor",
        .default = "background",
    ), levels = gene_labels_orderby))
```


## Labeling genes based on gene-list (directed)
1. Filter interactions based on source and target of interest
2. Only keep interactions that have at least one of the genes in the provided gene list 
3. Format pre-computed average expression and label genes based on whether they are part of an interaction of interest (see previous point 2)
4. Visualize scatter plot
```{r label_by_genes}
genes_oi <- c("L1CAM", "NRXN")
print(genes_oi)

res_genes_oi <- generate_scatter_plots(
    expr_df = df,
    detected_interactions_path = params$input_file,
    condition_oi = params$condition_oi,
    condition_varname = params$condition_varname,
    source_oi = params$source,
    target_oi = params$target,
    cutoff = params$cutoff,
    user_input_list = genes_oi,
    is_directed = TRUE
)
# Save plots
ggsave(plot = res_genes_oi$p_loess, filename = glue("{params$output_dir}/scatter__{params$condition_oi}__{source_target_oi_formatted}_w_LOESS__genes_oi__directed.pdf"), width = 12, height = 12)
ggsave(plot = res_genes_oi$p, filename = glue("{params$output_dir}/scatter__{params$condition_oi}__{source_target_oi_formatted}__genes_oi__directed.pdf"), width = 12, height = 12)


knitr::kable(res_genes_oi$interactions_oi)

# Visualize scatterplots in notebook
res_genes_oi$p_loess

res_genes_oi$p
```

## Labeling genes based on list of interactions (directed)
1. Obtain unique interactions from provided Excel file with unique interactions 
2. Filter interactions based on source and target of interest 
3. Only keep interactions that are part of the interactions of interest (point 1)
4. Format pre-computed average expression and label genes based on whether they are part of an interaction of interest (see previous point 2)
```{r label_by_interactions_oi}
interactions_list <- get_genes_from_unique_interactions(
    source_oi = params$source,
    target_oi = params$target,
    unique_interactions_path = params$unique_interactions,
    condition_varname = params$condition_varname,
    condition_oi = params$condition_oi,
    is_directed = TRUE,
    # Top_n = NULL -> all genes from unique interactions
    top_n = params$top_n
)
print(interactions_list)

res_interactions_oi <- generate_scatter_plots(
    expr_df = df,
    detected_interactions_path = params$input_file,
    condition_oi = params$condition_oi,
    condition_varname = params$condition_varname,
    source_oi = params$source,
    target_oi = params$target,
    cutoff = params$cutoff,
    user_input_list = interactions_list,
    is_directed = TRUE
)

# Save plots
ggsave(plot = res_interactions_oi$p_loess, filename = glue("{params$output_dir}/scatter__{params$condition_oi}__{source_target_oi_formatted}_w_LOESS__unique_interactions__directed.pdf"), width = 12, height = 12)
ggsave(plot = res_interactions_oi$p, filename = glue("{params$output_dir}/scatter__{params$condition_oi}__{source_target_oi_formatted}__unique_interactions__directed.pdf"), width = 12, height = 12)


knitr::kable(res_interactions_oi$interactions_oi)

# Visualize scatterplots in notebook
res_interactions_oi$p_loess

res_interactions_oi$p
```



## Labeling genes based on gene-list (undirected)
1. Filter interactions based on source and target of interest (undirected)
2. Only keep interactions that have at least one of the genes in the provided gene list 
3. Format pre-computed average expression and label genes based on whether they are part of an interaction of interest (see previous point 2)
4. Visualize scatter plot
```{r label_by_genes_undirected}
genes_oi <- c("L1CAM", "NRXN")
print(genes_oi)

res_genes_oi <- generate_scatter_plots(
    expr_df = df,
    detected_interactions_path = params$input_file,
    condition_oi = params$condition_oi,
    condition_varname = params$condition_varname,
    source_oi = params$source,
    target_oi = params$target,
    cutoff = params$cutoff,
    user_input_list = genes_oi,
    is_directed = FALSE
)
# Save plots
ggsave(plot = res_genes_oi$p_loess, filename = glue("{params$output_dir}/scatter__{params$condition_oi}__{source_target_oi_formatted}_w_LOESS__genes_oi__undirected.pdf"), width = 12, height = 12)
ggsave(plot = res_genes_oi$p, filename = glue("{params$output_dir}/scatter__{params$condition_oi}__{source_target_oi_formatted}__genes_oi__undirected.pdf"), width = 12, height = 12)


knitr::kable(res_genes_oi$interactions_oi)

# Visualize scatterplots in notebook
res_genes_oi$p_loess
res_genes_oi$p
```

## Labeling genes based on list of interactions (undirected)
1. Obtain unique interactions from provided Excel file with unique interactions 
2. Filter interactions based on source and target of interest 
3. Only keep interactions that are part of the interactions of interest (point 1)
4. Format pre-computed average expression and label genes based on whether they are part of an interaction of interest (see previous point 2)
```{r label_by_interactions_oi_undirected}
interactions_list <- get_genes_from_unique_interactions(
    source_oi = params$source,
    target_oi = params$target,
    unique_interactions_path = params$unique_interactions,
    condition_varname = params$condition_varname,
    condition_oi = params$condition_oi,
    is_directed = FALSE,
    # Top_n = NULL -> all genes from unique interactions
    top_n = params$top_n
)
print(interactions_list)

res_interactions_oi <- generate_scatter_plots(
    expr_df = df,
    detected_interactions_path = params$input_file,
    condition_oi = params$condition_oi,
    condition_varname = params$condition_varname,
    source_oi = params$source,
    target_oi = params$target,
    cutoff = params$cutoff,
    user_input_list = interactions_list,
    is_directed = FALSE
)

# Save plots
ggsave(plot = res_interactions_oi$p_loess, filename = glue("{params$output_dir}/scatter__{params$condition_oi}__{source_target_oi_formatted}_w_LOESS__unique_interactions__undirected.pdf"), width = 12, height = 12)
ggsave(plot = res_interactions_oi$p, filename = glue("{params$output_dir}/scatter__{params$condition_oi}__{source_target_oi_formatted}__unique_interactions__undirected.pdf"), width = 12, height = 12)


knitr::kable(res_interactions_oi$interactions_oi)

# Visualize scatterplots in notebook
res_interactions_oi$p_loess

res_interactions_oi$p
```

# Session Info
```{r session_info}
sessioninfo::session_info()
```